File: JavaScript/ExampleTests/common/ShippingTyped.js
import { ShippingString } from "./ShippingString.js";
import * as _json from "./json.js";

export class ShippingTyped {
  constructor(totalPrice = "", shippingCost = "", notes = "") {
    this.totalPrice = totalPrice;
    this.shippingCost = shippingCost;
    this.notes = notes;
  }

  static fromStringObj(s) {
    return new ShippingTyped(
      new Dollar(s.totalPrice),
      new Dollar(s.shippingCost),
      s.notes
    );
  }

  toJsonValue() {
    return {
      totalPrice: this.totalPrice == null ? null : String(this.totalPrice),
      shippingCost: this.shippingCost == null ? null : String(this.shippingCost),
      notes: this.notes,
    };
  }

  toJSON() { return _json.stringify(this.toJsonValue()); }

  static fromJsonValue(m) {
    return new ShippingTyped(
      new Dollar(_json.asString(_json.require(m, "totalPrice"), "totalPrice")),
      new Dollar(_json.asString(_json.require(m, "shippingCost"), "shippingCost")),
      _json.asString(_json.require(m, "notes"), "notes")
    );
  }

  static fromJSON(text) { return ShippingTyped.fromJsonValue(_json.parse(text)); }

  static toJSONList(items) {
    return _json.stringify(items.map((item) => item.toJsonValue()));
  }

  static fromJSONList(text) {
    const raw = _json.asArray(_json.parse(text), "ShippingTyped");
    return raw.map((e) => ShippingTyped.fromJsonValue(e));
  }
}

File: Rust/ExampleTests/common/shipping_typed.rs
#![allow(dead_code, unused_imports, unused_variables)]

use super::json;
use super::shipping_string::ShippingString;

#[derive(Debug, Clone, Default)]
pub struct ShippingTyped {
    pub total_price: Dollar,
    pub shipping_cost: Dollar,
    pub notes: String,
}

impl ShippingTyped {
    pub fn from_str_struct(s: &ShippingString) -> Self {
        Self {
            total_price: Dollar::from(s.total_price.clone()),
            shipping_cost: Dollar::from(s.shipping_cost.clone()),
            notes: s.notes.clone(),
        }
    }

    pub fn to_json_value(&self) -> json::Value {
        json::Value::Object(vec![
            ("total_price".to_string(), json::Value::Str(self.total_price.to_string())),
            ("shipping_cost".to_string(), json::Value::Str(self.shipping_cost.to_string())),
            ("notes".to_string(), json::Value::Str(self.notes.clone())),
        ])
    }

    pub fn to_json(&self) -> String {
        json::write(&self.to_json_value())
    }

    pub fn from_json_value(v: &json::Value) -> json::JsonResult<Self> {
        Ok(Self {
            total_price: Dollar::from(json::as_string(json::require(v, "total_price")?, "total_price")?),
            shipping_cost: Dollar::from(json::as_string(json::require(v, "shipping_cost")?, "shipping_cost")?),
            notes: json::as_string(json::require(v, "notes")?, "notes")?,
        })
    }

    pub fn from_json(text: &str) -> json::JsonResult<Self> {
        Self::from_json_value(&json::parse(text)?)
    }

    pub fn to_json_list(list: &[ShippingTyped]) -> String {
        json::write(&json::Value::Array(
            list.iter().map(|item| item.to_json_value()).collect(),
        ))
    }

    pub fn from_json_list(text: &str) -> json::JsonResult<Vec<ShippingTyped>> {
        let root = json::parse(text)?;
        let items = json::as_array(&root, "ShippingTyped")?;
        let mut result = Vec::with_capacity(items.len());
        for e in items {
            result.push(Self::from_json_value(e)?);
        }
        Ok(result)
    }
}
